wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wgtun provides a Device implementing the tun.Device interface
+// with additional features as required by wgengine.
+package wgtun
+
+import (
+	"errors"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/tun"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	readMaxSize   = device.MaxMessageSize
+	readOffset    = device.MessageTransportHeaderSize
+	MaxPacketSize = device.MaxContentSize
+)
+
+var (
+	ErrFiltered     = errors.New("packet dropped by filter")
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// Device wraps a tun.Device from wireguard-go,
```

Comment:
Open-ended question for you: what's a better name for this?

I see that you're copying the tun.Device name from wireguard-go, and given that name this name makes sense. But that upstream name is confusing, there is both a tun.Device and a device.Device. Some day I'd like to come up with something less confusing there.

Should wireguard-go's tun.Device be tun.Tun? If so, that would make this wgtun.Tun. Is that a reasonable name here?

Reply:
That's a good point, `wgtun.Tun` (I guess `wgtun.TUN`, following the conventions?) is less confusing than `wgtun.Device` to me.

Reply:
At the same time, it seems stuttery.

Reply:
I've renamed it to `wgtun.TUN` anyway. I preferred the old scheme, but maybe that's just because I got used to it in the process of writing this. I'd love for others to weigh in.